guava-tests/test/com/google/common/hash/Murmur3Hash32Test.java

```diff
@@ -51,27 +52,37 @@ public void testKnownStringInputs() {
         -528633700, murmur3_32().hashUnencodedChars("The quick brown fox jumps over the lazy dog"));
   }
 
-  public void testKnownUtf8StringInputs() {
-    assertHash(0, murmur3_32().hashString("", Charsets.UTF_8));
-    assertHash(0xcfbda5d1, murmur3_32().hashString("k", Charsets.UTF_8));
-    assertHash(0xa167dbf3, murmur3_32().hashString("hell", Charsets.UTF_8));
-    assertHash(0x248bfa47, murmur3_32().hashString("hello", Charsets.UTF_8));
-    assertHash(0x3d41b97c, murmur3_32().hashString("http://www.google.com/", Charsets.UTF_8));
-    assertHash(
-        0x2e4ff723,
-        murmur3_32().hashString("The quick brown fox jumps over the lazy dog", Charsets.UTF_8));
-    assertHash(0xfc5ba834, murmur3_32().hashString("毎月１日,毎週月曜日", Charsets.UTF_8));
+  public void testKnownEncodedStringInputs() {
+    assertStringHash(0, "", Charsets.UTF_8);
+    assertStringHash(0xcfbda5d1, "k", Charsets.UTF_8);
+    assertStringHash(0xa167dbf3, "hell", Charsets.UTF_8);
+    assertStringHash(0x248bfa47, "hello", Charsets.UTF_8);
+    assertStringHash(0x3d41b97c, "http://www.google.com/", Charsets.UTF_8);
+    assertStringHash(0x2e4ff723, "The quick brown fox jumps over the lazy dog", Charsets.UTF_8);
+    assertStringHash(0xb5a4be05, "ABCDefGHI\u0799", Charsets.UTF_8);
+    assertStringHash(0xfc5ba834, "毎月１日,毎週月曜日", Charsets.UTF_8);
+    assertStringHash(0x8a5c3699, "surrogate pair: \uD83D\uDCB0", Charsets.UTF_8);
+
+    assertStringHash(0, "", Charsets.UTF_16);
+    assertStringHash(0xae9d4799, "k", Charsets.UTF_16);
```

Comment:
There's a subtle problem with this. (At least, it took me a while to figure it out.) A byte array from the `UTF_16` encoding starts with a two-byte Byte Order Mark (BOM) which is either `fe ff` or `ff fe` and indicates the endianness that the remaining bytes will use. It's not specified which endianness a given Java platform will use, so it's not correct to hash a string using the `UTF_16` encoding if you want the result to be portable. I discovered this because some non-public tests were using the opposite endianness and failing. I rewrote these test cases to use `UTF_16LE` and updated the hashcodes. (You don't need to do that here but if we go ahead with this change then the modified test will be in the version that we use.)

Reply:
@eamonnmcmanus indeed, this isn't portable. Thanks for catching this.
UTF-16LE is as good, as the only point of this test coverage is to exercise the code path that was not optimized for UTF-8.

If there is any value in changing anything in this PR, i am happy to apply changes.

Reply:
I saw a reference to this conversation, and I dug up [Android issue 37074504](https://issuetracker.google.com/issues/37074504) ([fix](https://android-review.googlesource.com/c/platform/libcore/+/437155)), last seen [as "https://code.google.com/p/android/issues/detail?id=196848" in `ByteStreamsTest`](https://github.com/google/guava/blob/d32a049827c0dd588b51c7bbb98d54eb7384a602/guava-tests/test/com/google/common/io/ByteStreamsTest.java#L383). I think what Éamonn is seeing is a bug in the very old version of Android that we test with.

UTF-16LE is still the right pragmatic fix -- and probably the one I should have employed in `ByteStreamsTest`!

Reply:
Done, changed to UTF-16LE (and updated the expected hashes accordingly).